File: chrome/browser/android/ssl/mock_cert_verifier_rule_android.cc
// Copyright 2018 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/android/ssl/mock_cert_verifier_rule_android.h"

#include "base/command_line.h"

// Must come after all headers that specialize FromJniType() / ToJniType().
#include "chrome/android/test_support_jni_headers/MockCertVerifierRuleAndroid_jni.h"

jlong JNI_MockCertVerifierRuleAndroid_Init(JNIEnv* env) {
  return reinterpret_cast<intptr_t>(new MockCertVerifierRuleAndroid());
}

MockCertVerifierRuleAndroid::MockCertVerifierRuleAndroid() = default;

void MockCertVerifierRuleAndroid::SetResult(JNIEnv* env, int result) {
  mock_cert_verifier_.mock_cert_verifier()->set_default_result(result);
}

void MockCertVerifierRuleAndroid::SetUp(JNIEnv* env) {
  mock_cert_verifier_.SetUpCommandLine(base::CommandLine::ForCurrentProcess());
  mock_cert_verifier_.SetUpInProcessBrowserTestFixture();
}

void MockCertVerifierRuleAndroid::TearDown(JNIEnv* env) {
  mock_cert_verifier_.TearDownInProcessBrowserTestFixture();
}

File: chrome/android/javatests/src/org/chromium/chrome/browser/test/MockCertVerifierRuleAndroid.java
// Copyright 2018 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.test;

import org.jni_zero.NativeMethods;
import org.junit.rules.ExternalResource;

import org.chromium.content_public.browser.test.NativeLibraryTestUtils;

/** JUnit test rule which enables tests to force certificate verification results. */
public class MockCertVerifierRuleAndroid extends ExternalResource {

    private long mNativePtr;

    // Certificate verification result to force.
    private int mResult;

    public MockCertVerifierRuleAndroid(int result) {
        mResult = result;
    }

    @Override
    protected void before() {
        NativeLibraryTestUtils.loadNativeLibraryNoBrowserProcess();
        mNativePtr = MockCertVerifierRuleAndroidJni.get().init();
        MockCertVerifierRuleAndroidJni.get().setResult(mNativePtr, mResult);
        MockCertVerifierRuleAndroidJni.get().setUp(mNativePtr);
    }

    public void setResult(int result) {
        mResult = result;
        if (mNativePtr != 0) {
            MockCertVerifierRuleAndroidJni.get().setResult(mNativePtr, result);
        }
    }

    @Override
    protected void after() {
        MockCertVerifierRuleAndroidJni.get().tearDown(mNativePtr);
        mNativePtr = 0;
    }

    @NativeMethods
    interface Natives {
        long init();

        void setUp(long nativeMockCertVerifierRuleAndroid);

        void setResult(long nativeMockCertVerifierRuleAndroid, int result);

        void tearDown(long nativeMockCertVerifierRuleAndroid);
    }
}
